### PR TITLE
fix: observe grpc client

### PR DESCRIPTION
**Pull request by gatici**
This PR aims to solve GPRC client connection issues which appears when GPRC server (webconsole) is restarted.

If the Webconsole is restarted, other modules loses the GRPC connection to Webconsole
They tries to connect again for hours with no success.

2024-09-26T11:32:10.014Z [smf] 2024-09-26T11:32:10Z [ERRO][Config5g][GRPC] Connectivity status idle, trying to connect again

NF observes the GRPC client availability and connection status in an infinite loop.
If the connection is ready it does nothing. If GRPC client does not exist, it creates new client. If GRPC client connection is not ready, it closes the existing client and creates another client.
Only one GRPC client exists for any time slot.

### CODE REVIEW

**Review by gab-arrobo — COMMENTED** [review]
I see that changes are not consistent across the different NFs (in some NFs the `updateConfig` is in `service/init.go` like this in NF/PR and in other NFs, the same function goes in `factory/factory.go`). Would this be a good time to make the change consistent across all NFs?

File: service/init.go [hunk 1]
```diff
@@ -107,11 +107,13 @@ func (udm *UDM) Initialize(c *cli.Context) error {
...
-		commChannel := client.ConfigWatcher(factory.UdmConfig.Configuration.WebuiUri)
-		go udm.updateConfig(commChannel)
+		client, err := grpcClient.ConnectToConfigServer(factory.UdmConfig.Configuration.WebuiUri)
+		if err != nil {
+			go updateConfig(client, udm)
+		}
+		return err
 	} else {
 		go func() {
 			initLog.Infoln("Use helm chart config ")
```
[the remark this answers is not in the document]
Comment (gatici) [reply service/init.go]: Done

**Comment by gab-arrobo** [comment]
@gatici did you have a chance to test this PR? I was trying to run an E2E test using `gnbsim` but my test fails when using your changes from this PR.

**Comment by gatici** [comment]
> @gatici did you have a chance to test this PR? I was trying to run an E2E test using `gnbsim` but my test fails when using your changes from this PR.

I am working on the test part. I will keep you updated.

File: service/init.go [hunk 2]
```diff
@@ -101,21 +102,61 @@ func (udm *UDM) Initialize(c *cli.Context) error {
...
+			stream, err = client.CheckGrpcConnectivity()
+			if err != nil {
+				logger.InitLog.Errorf("%v", err)
+			}
+			if stream == nil {
+				time.Sleep(time.Second * 30)
+				continue
+			}
+			if client.GetConfigClientConn().GetState() != connectivity.Ready {
+				err := client.GetConfigClientConn().Close()
```
Comment (gab-arrobo) [inline service/init.go:139]:
```suggestion
				err = client.GetConfigClientConn().Close()
```

Comment (gatici) [reply service/init.go]: Done

**Comment by gatici** [comment]
@gab-arrobo  Updated config5g is used and E2E tests are passed successfully.

**Comment by gatici** [comment]
The implementations are done under service/init.go Initialize method in all the NF's.

File: service/init.go [hunk 3]
```diff
@@ -101,21 +101,71 @@ func (udm *UDM) Initialize(c *cli.Context) error {
...
-	if roc == "true" {
+	if os.Getenv("MANAGED_BY_CONFIG_POD") == "true" {
 		logger.InitLog.Infoln("MANAGED_BY_CONFIG_POD is true")
-		commChannel := client.ConfigWatcher(factory.UdmConfig.Configuration.WebuiUri)
-		go udm.updateConfig(commChannel)
+		go manageGrpcClient(factory.UdmConfig.Configuration.WebuiUri, udm)
 	} else {
 		go func() {
-			logger.InitLog.Infoln("use helm chart config")
+			logger.InitLog.Infoln("use helm chart config ")
```
Comment (gab-arrobo) [inline service/init.go:109]: Revert this change
```suggestion
			logger.InitLog.Infoln("use helm chart config")
```

**Review by gab-arrobo — APPROVED** [review]
